repository.go

```diff
@@ -1223,3 +1223,85 @@ func (r *Repository) createNewObjectPack(cfg *RepackConfig) (h plumbing.Hash, er
 
 	return h, err
 }
+
+type Describe struct {
+	// Reference being described
+	Reference *plumbing.Reference
+	// Tag of the describe object
+	Tag *plumbing.Reference
+	// Distance to the tag object in commits
+	Distance int
+	// Dirty string to append
+	Dirty string
+	// Use <Abbrev> digits to display SHA-ls
+	Abbrev int
+}
+
+func (d *Describe) String() string {
+	var s []string
+
+	if d.Tag != nil{
+		s = append(s, d.Tag.Name().Short())
+	}
+	if d.Distance > 0 {
+		s = append(s, fmt.Sprint(d.Distance))
+	}
+	s = append(s, d.Reference.Hash().String()[0:d.Abbrev])
+	if d.Dirty != "" {
+		s = append(s, d.Dirty)
+	}
+
+	return strings.Join(s, "-")
+}
+
+// Describe just like the `git describe` command will return a Describe struct for the hash passed.
+// Describe struct implements String interface so it can be easily printed out.
+func (r *Repository) Describe(ref *plumbing.Reference, opts *DescribeOptions) (*Describe, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	// Fetch the reference log
+	commitIterator, err := r.Log(&LogOptions{
+		From:  ref.Hash(),
+		Order: LogOrderCommitterTime,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Tag map for rapid (?) query
+	tagIterator, err := r.Tags()
```

Comment:
Maybe this can be implemented at `Repository` with a function `Tag(referenceName)`, or a helper function

Reply:
Something like [TagObject](https://github.com/src-d/go-git/blob/master/repository.go#L939) but for lightweight tags will be ideal. 
Then we just need to walk down the log checking for annotated tags or lw-tags deppending on the `--tag` option always preferring annotated ones as described in the [git documentation](https://git-scm.com/docs/git-describe#_search_strategy). This search algorithm is nicely described doing a `git describe --tags --debug`:
```bash
$ git describe --tags --debug
searching to describe HEAD
 lightweight        2 v4.3.1
 lightweight       12 v4.3.0
 lightweight       18 v4.2.1
 lightweight       35 v4.2.0
 lightweight       56 v4.1.1
 lightweight       77 v4.1.0
 lightweight       90 v4.0.0
 lightweight      271 v4.0.0-rc15
 lightweight      282 v4.0.0-rc14
 lightweight      330 v4.0.0-rc13
traversed 393 commits
more than 10 tags found; listed 10 most recent
gave up search at d3c7400c39f86a4c59340c7a9cda8497186e00fc
v4.3.1-2-g87cc819
```

I did not dig into the details of `TagObject`. I'll check.